### PR TITLE
METRON-1700: Create REST endpoint to get job configuration

### PR DESCRIPTION
**Pull request by merrimanr**
## Contributor Comments
This PR adds a REST endpoint that returns the job configuration for a submitted job.  The endpoint returns a static list of properties that can be serialized and are useful for frontend applications.  Changes outside of REST were trivial because the PcapJob class already stores the configuration.

### Changes Included
- Added a method to the Statusable interface that returns a configuration map
- Added a method to the PcapService that filters the Statusable configuration and returns a subset of properties
- Update unit and integration tests

### Testing
This has been tested in full dev.  Testing instructions are as follows:

1. Spin up full dev and put pcap data in the `/apps/metron/pcap/input` HDFS directory
2. Submit a pcap query:
```
curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' -d '{}' 'http://node1:8082/api/v1/pcap/fixed'
```
A job status should be returned as usual.

3. Now the configuration can be retrieved with the get configuration endpoint:
```
curl -X GET --header 'Accept: application/json' 'http://node1:8082/api/v1/pcap/job_1532356159290_0052/configuration'
```
A subset of properties, including those in the pcap query:
```
{
  "finalOutputPath": "/apps/metron/pcap/output",
  "ipSrcPort": "",
  "includeReverse": false,
  "startTimeMs": 1449010818000,
  "endTimeMs": 1532642418000,
  "protocol": "",
  "ipDstAddr": "",
  "basePath": "/apps/metron/pcap/input",
  "packetFilter": "",
  "ipSrcAddr": "",
  "ipDstPort": "",
  "numReducers": 10,
}
```

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: metron-interface/metron-rest/src/main/java/org/apache/metron/rest/service/impl/PcapServiceImpl.java [hunk 1]
```diff
@@ -199,6 +208,37 @@ public InputStream getRawPcap(String username, String jobId, Integer page) throw
     return inputStream;
   }
 
+  @Override
+  public Map<String, Object> getConfiguration(String username, String jobId) throws RestException {
+    Map<String, Object> configuration = new HashMap<>();
+    try {
+      Map<String, Object> jobConfiguration = jobManager.getJob(username, jobId).getConfiguration();
```
Comment (nickwallen) [inline metron-interface/metron-rest/src/main/java/org/apache/metron/rest/service/impl/PcapServiceImpl.java]: Will `jobManager.getJob(username, jobId)` never return null?  I see null checks in other places where similar calls are made.

File: metron-interface/metron-rest/src/main/java/org/apache/metron/rest/service/PcapService.java [hunk 2]
```diff
@@ -43,5 +44,5 @@
 
   InputStream getRawPcap(String username, String jobId, Integer page) throws RestException;
 
-
+  Map<String, Object> getConfiguration(String username, String jobId) throws RestException;
```
Comment (nickwallen) [inline metron-interface/metron-rest/src/main/java/org/apache/metron/rest/service/PcapService.java]: Would be nice to see some javadocs here, but that doesn't seem unique to this PR.  Maybe in another PR we can get some javadocs in here.

File: metron-interface/metron-rest/src/main/java/org/apache/metron/rest/service/impl/PcapServiceImpl.java [hunk 3]
```diff
@@ -199,6 +208,37 @@ public InputStream getRawPcap(String username, String jobId, Integer page) throw
...
+        configuration.put(FixedPcapOptions.IP_SRC_PORT.getKey(), FixedPcapOptions.IP_SRC_PORT.get(jobConfiguration, String.class));
+        configuration.put(FixedPcapOptions.IP_DST_PORT.getKey(), FixedPcapOptions.IP_DST_PORT.get(jobConfiguration, String.class));
+        configuration.put(FixedPcapOptions.PROTOCOL.getKey(), FixedPcapOptions.PROTOCOL.get(jobConfiguration, String.class));
+        configuration.put(FixedPcapOptions.PACKET_FILTER.getKey(), FixedPcapOptions.PACKET_FILTER.get(jobConfiguration, String.class));
+        configuration.put(FixedPcapOptions.INCLUDE_REVERSE.getKey(), FixedPcapOptions.INCLUDE_REVERSE.get(jobConfiguration, String.class));
+      } else {
+        configuration.put(QueryPcapOptions.QUERY.getKey(), QueryPcapOptions.QUERY.get(jobConfiguration, String.class));
+      }
+    } catch (JobNotFoundException e) {
+      // do nothing and return null pcapStatus
```
Comment (nickwallen) [inline metron-interface/metron-rest/src/main/java/org/apache/metron/rest/service/impl/PcapServiceImpl.java]: Is this comment correct?  Seems like we return an empty configuration map (assuming the call to getJob does not return null).

File: metron-interface/metron-rest/src/test/java/org/apache/metron/rest/controller/PcapControllerIntegrationTest.java [hunk 4]
```diff
@@ -420,5 +420,55 @@ public void testRawDownload() throws Exception {
...
+            .andExpect(jsonPath("$.ipDstAddr").value("192.168.1.1"))
+            .andExpect(jsonPath("$.ipSrcPort").value("2000"))
+            .andExpect(jsonPath("$.ipDstPort").value("1000"))
+            .andExpect(jsonPath("$.protocol").value("TCP"))
+            .andExpect(jsonPath("$.packetFilter").value("filter"))
+            .andExpect(jsonPath("$.includeReverse").value("true"));
+  }
 
+  @Test
+  public void testGeQueryFilterConfiguration() throws Exception {
```
Comment (nickwallen) [inline metron-interface/metron-rest/src/test/java/org/apache/metron/rest/controller/PcapControllerIntegrationTest.java]: typo... testGeQueryFilterConfiguration != testGetQueryFilterConfiguration

File: metron-interface/metron-rest/src/test/java/org/apache/metron/rest/service/impl/PcapServiceImplTest.java [hunk 5]
```diff
@@ -639,4 +642,78 @@ public void getRawShouldThrowException() throws Exception {
...
+    Assert.assertEquals("ip_dst_addr", FixedPcapOptions.IP_DST_ADDR.get(configuration, String.class));
+    Assert.assertEquals(1000, FixedPcapOptions.IP_SRC_PORT.get(configuration, Integer.class).intValue());
+    Assert.assertEquals(2000, FixedPcapOptions.IP_DST_PORT.get(configuration, Integer.class).intValue());
+    Assert.assertEquals("tcp", FixedPcapOptions.PROTOCOL.get(configuration, String.class));
+    Assert.assertEquals("filter", FixedPcapOptions.PACKET_FILTER.get(configuration, String.class));
+    Assert.assertEquals(true, FixedPcapOptions.INCLUDE_REVERSE.get(configuration, Boolean.class));
+  }
+
+  @Test
+  public void getConfigurationShouldProperlyReturnQueryFilterConfiguration() throws Exception {
```
Comment (nickwallen) [inline metron-interface/metron-rest/src/test/java/org/apache/metron/rest/service/impl/PcapServiceImplTest.java]: Is there a test case for trying to get the configuration of a job ID that does not exist?

File: metron-interface/metron-rest/src/main/java/org/apache/metron/rest/service/impl/PcapServiceImpl.java (hunk 3 above)
Comment (merrimanr) [reply metron-interface/metron-rest/src/main/java/org/apache/metron/rest/service/impl/PcapServiceImpl.java]: The pattern is used in other places in the class and the comment was copied.  I updated it to say an empty map is returned.

Comment (ottobackwards) [reply metron-interface/metron-rest/src/main/java/org/apache/metron/rest/service/impl/PcapServiceImpl.java]: We should log here.

**Comment by merrimanr** [comment]
I added 2 commits, one for @nickwallen's feedback and one for @ottobackwards's feedback.

File: metron-interface/metron-rest/src/test/java/org/apache/metron/rest/service/impl/PcapServiceImplTest.java (hunk 5 above)
Comment (merrimanr) [reply metron-interface/metron-rest/src/test/java/org/apache/metron/rest/service/impl/PcapServiceImplTest.java]: I added a couple test cases for exception handling.

File: metron-interface/metron-rest/src/test/java/org/apache/metron/rest/controller/PcapControllerIntegrationTest.java (hunk 4 above)
Comment (merrimanr) [reply metron-interface/metron-rest/src/test/java/org/apache/metron/rest/controller/PcapControllerIntegrationTest.java]: Done

File: metron-interface/metron-rest/src/main/java/org/apache/metron/rest/service/impl/PcapServiceImpl.java (hunk 3 above)
Comment (merrimanr) [reply metron-interface/metron-rest/src/main/java/org/apache/metron/rest/service/impl/PcapServiceImpl.java]: I replaced the empty catch blocks with a log warning.  Will that work?

File: metron-interface/metron-rest/src/main/java/org/apache/metron/rest/service/PcapService.java (hunk 2 above)
Comment (merrimanr) [reply metron-interface/metron-rest/src/main/java/org/apache/metron/rest/service/PcapService.java]: I added javadocs for this service.

File: metron-interface/metron-rest/README.md [hunk 6]
```diff
@@ -556,6 +558,14 @@ Request and Response objects are JSON formatted.  The JSON schemas are available
     * jobId - Job ID of submitted job
   * Returns:
     * 200 - Kills passed job.
+    
+### `GET /api/v1/pcap/{jobId}/configuration`
```
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-interface/metron-rest/README.md]: I agree.  Done.

**Comment by nickwallen** [comment]
+1 Looks good.  Thanks for the updates.

**Pull request closed without merging**